scripts/generate_matrix.py

```diff
@@ -0,0 +1,227 @@
+        if version_str.count(".") == 1:
+            version_str += ".0"
+        return cls(*map(int, version_str.strip().split(".")))
+
+
+def get_minimum_python_version(package: str) -> Version:
+    """
+    Fetch the minimum Python version specified in pyproject.toml.
+
+    Returns:
+        str: The minimum required version of the package.
+
+    Examples:
+        >>> get_minimum_python_version()  # doctest: +ELLIPSIS
+        Version(major=3, ...)
+    """
+    raw_version = importlib.metadata.metadata(package)["Requires-Python"]
+    if "<" in raw_version:
+        raise NotImplementedError("Version specifier with upper bound not yet supported!")
+
+    return Version.from_str(raw_version.replace(">=", "").replace("~=", ""))
+
+
+def get_minimum_version(package: str, dependency: str) -> Version:
+    """
+    Fetch the minimum dependency version required by the given package.
+
+    Args:
+        package (str): The package name ('scipy' or package name).
+        dependency (str): The dependency name ('numpy' or dependency name).
+
+    Returns:
+        str: The minimum required version of the package.
+
+    Examples:
+        >>> get_minimum_version("scipy", "numpy")  # doctest: +ELLIPSIS
+        Version(major=1, ...)
+    """
+    np_minimum_dep = next(
+        req for req in importlib.metadata.requires(package) if req.startswith(dependency) and " extra " not in req
+    )
+    np_minimum_dep = next(ver for ver in np_minimum_dep.split(",") if ">" in ver)
+    return Version.from_str(np_minimum_dep.replace(dependency, "").replace(">=", "").replace(">", ""))
+
+
+def get_python_versions(
+    min_version: Version | None = None,
+    max_version: Version | None = None,
+) -> list[Version]:
+    """
+    Fetch the list of available Python versions from the GitHub Actions Python Versions Manifest.
+
+    Args:
+        min_version (Version): The minimum version to include.
+        max_version (Version): The maximum version to include.
+        include_prerelease (bool): Whether to include pre-release versions.
+
+    Returns:
+        list[Version]: A list of available Python versions.
+
+    Raises:
+        urllib.error.URLError: If fetching data fails.
+
+    Examples:
+        >>> versions = get_python_versions()
+        >>> versions[0] >= Version(3, 6, 0)
+        True
+    """
+    data = fetch_json("https://raw.githubusercontent.com/actions/python-versions/main/versions-manifest.json")
+
+    versions = {}
+
+    for release in data[::-1]:
+        try:
+            version = Version.from_str(release["version"])
+        except ValueError:
+            # Skip pre-release versions
+            continue
+
+        if min_version and version < min_version:
+            continue
+        if max_version and version > max_version:
+            continue
+
+        versions[version.minor_version] = version
+
+    return sorted(versions.values())
+
+
+@lru_cache(maxsize=CACHE_SIZE)
+def fetch_json(url: str) -> dict | list:
```

Comment:
`dict` and `list` are missing their type arguments

Reply:
I've added an any, no point in making these type hints more complicated for now :)

Reply:
not needed here, but in case you encounter something like this again, then there's [`optype.json`](https://github.com/jorenham/optype?tab=readme-ov-file#optypejson) that could help with that